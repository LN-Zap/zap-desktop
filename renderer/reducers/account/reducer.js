import createReducer from '@zap/utils/createReducer'
import { getIntl } from '@zap/i18n'
import { mainLog } from '@zap/utils/log'
import waitForIpcEvent from '@zap/utils/waitForIpc'
import { closeDialog } from 'reducers/modal'
import { showNotification } from 'reducers/notification'
import accountSelectors from './selectors'
import messages from './messages'
import * as constants from './constants'

const {
  INIT_ACCOUNT,
  INIT_ACCOUNT_SUCCESS,
  INIT_ACCOUNT_FAILURE,
  LOGIN,
  LOGIN_SUCCESS,
  LOGIN_FAILURE,
  LOGIN_CLEAR_ERROR,
  CHANGE_PASSWORD_DIALOG_ID,
  PASSWORD_PROMPT_DIALOG_ID,
  PASSWORD_SET_DIALOG_ID,
  SET_IS_PASSWORD_ENABLED,
  LOGIN_NOT_ALLOWED,
} = constants

// ------------------------------------
// Initial State
// ------------------------------------

export const initialState = {
  isAccountLoading: false,
  isAccountLoaded: false,
  initAccountError: null,
  isLoggingIn: false,
  loginError: null,
  isLoggedIn: false,
  isPasswordEnabled: null,
}

// ------------------------------------
// Actions
// ------------------------------------

/**
 * initAccount - Fetch the current account info from the database and save into the store.
 * Should be called once when the app first loads.
 *
 * @returns {Function} Thunk
 */
export const initAccount = () => async dispatch => {
  dispatch({ type: INIT_ACCOUNT })
  try {
    // Auto login user if password feature is disabled.
    const isAccountPasswordEnabled = await dispatch(checkAccountPasswordEnabled())
    if (!isAccountPasswordEnabled) {
      dispatch({ type: LOGIN_SUCCESS })
    }
    dispatch({ type: INIT_ACCOUNT_SUCCESS })
  } catch (e) {
    mainLog.warn('checkAccountPasswordEnabled error: %o', e)
    dispatch({ type: LOGIN_FAILURE, error: LOGIN_NOT_ALLOWED })
  }
}

/**
 * setIsPasswordEnabled - Whether password is set for the account.
 *
 * @param {string} value value
 * @returns {object} Action
 */
const setIsPasswordEnabled = value => ({
  type: SET_IS_PASSWORD_ENABLED,
  value,
})

/**
 * setPassword - Updates wallet password.
 *
 * @param {string} password new password
 * @returns {Function} Thunk
 */
const setPassword = password => async dispatch => {
  const { sha256digest } = window.Zap
  dispatch(waitForIpcEvent('setPassword', { value: await sha256digest(password) }))
}

/**
 * checkAccountPasswordEnabled - Checks whether app password is set via checking secure storage.
 * Dispatches setIsPasswordEnabled on completion.
 *
 * @returns {Function} Thunk
 */
const checkAccountPasswordEnabled = () => async dispatch => {
  const { value } = await dispatch(waitForIpcEvent('hasPassword'))
  dispatch(setIsPasswordEnabled(value))
  return value
}

/**
 * changePassword - Changes existing password.
 *
 * @param {object} params password params
 * @param {string} params.newPassword new password
 * @param {string} params.oldPassword old password
 * @returns {Function} Thunk
 */
export const changePassword = ({ newPassword, oldPassword }) => async dispatch => {
  try {
    const intl = getIntl()
    await dispatch(requirePassword(oldPassword))
    await dispatch(setPassword(newPassword))
    dispatch(closeDialog(CHANGE_PASSWORD_DIALOG_ID))
    dispatch(showNotification(intl.formatMessage(messages.account_password_updated)))
  } catch (error) {
    dispatch({ type: LOGIN_FAILURE, error: error.message })
  }
}

/**
 * enablePassword - Enables app-wide password protection.
 *
 * @param {string} password to be used further on.
 * @returns {Function} Thunk
 */
export const enablePassword = ({ password }) => async dispatch => {
  try {
    const intl = getIntl()
    await dispatch(setPassword(password))
    dispatch(closeDialog(PASSWORD_SET_DIALOG_ID))
    dispatch(setIsPasswordEnabled(true))
    dispatch(showNotification(intl.formatMessage(messages.account_password_enabled)))
  } catch (error) {
    dispatch({ type: LOGIN_FAILURE, error: error.message })
  }
}

/**
 * disablePassword - Disables app-wide password protection.
 *
 * @param {string} password current password.
 * @returns {Function} Thunk
 */
export const disablePassword = ({ password }) => async dispatch => {
  try {
    const intl = getIntl()
    await dispatch(requirePassword(password))
    await dispatch(waitForIpcEvent('deletePassword'))
    dispatch(setIsPasswordEnabled(false))
    dispatch(closeDialog(PASSWORD_PROMPT_DIALOG_ID))
    dispatch(showNotification(intl.formatMessage(messages.account_password_disabled)))
  } catch (error) {
    dispatch({ type: LOGIN_FAILURE, error: error.message })
  }
}

/**
 * requirePassword - Password protect routine. Should be placed before protected code.
 *
 * @param {string} password current password.
 * @returns {Promise} Promise that fulfills after login attempt (either successful or not)
 */
const requirePassword = password => async dispatch => {
  const { sha256digest } = window.Zap
  const { password: hash } = await dispatch(waitForIpcEvent('getPassword'))
  const passwordHash = await sha256digest(password)
  // compare hash received from the main thread to a hash of a password provided
  if (hash === passwordHash) {
    return true
  }
  throw new Error(getIntl().formatMessage(messages.account_invalid_password))
}

/**
 * login - Perform account login.
 *
 * @param {string} password Password
 * @returns {Function} Thunk
 */
export const login = password => async dispatch => {
  try {
    dispatch({ type: LOGIN, password })
    await dispatch(requirePassword(password))
    dispatch({ type: LOGIN_SUCCESS })
  } catch (error) {
    dispatch({ type: LOGIN_FAILURE, error: error.message })
  }
}

/**
 * clearLoginError - Clear login user error.
 *
 * @returns {Function} Thunk
 */
export const clearLoginError = () => (dispatch, getState) => {
  if (accountSelectors.loginError(getState())) {
    dispatch({
      type: 'LOGIN_CLEAR_ERROR',
    })
  }
}

// ------------------------------------
// Action Handlers
// ------------------------------------

const ACTION_HANDLERS = {
  [INIT_ACCOUNT]: state => {
    state.isAccountLoading = true
  },
  [INIT_ACCOUNT_SUCCESS]: state => {
    state.isAccountLoading = false
    state.isAccountLoaded = true
    state.initAccountError = null
  },
  [INIT_ACCOUNT_FAILURE]: (state, { error }) => {
    state.isAccountLoading = false
    state.initAccountError = error
  },

  [LOGIN]: state => {
    state.isLoggingIn = true
  },
  [LOGIN_SUCCESS]: state => {
    state.isLoggingIn = false
    state.isLoggedIn = true
    state.loginError = null
  },
  [LOGIN_FAILURE]: (state, { error }) => {
    state.isLoggingIn = false
    state.loginError = error
  },
  [LOGIN_CLEAR_ERROR]: state => {
    state.loginError = null
  },
  [SET_IS_PASSWORD_ENABLED]: (state, { value }) => {
    state.isPasswordEnabled = value
  },
}

// ------------------------------------
// Reducer
// ------------------------------------

export default createReducer(initialState, ACTION_HANDLERS)
