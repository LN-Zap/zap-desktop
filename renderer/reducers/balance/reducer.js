import Coin from '@zap/utils/coin'
import createReducer from '@zap/utils/createReducer'
import { grpc } from 'workers'

import * as constants from './constants'

const { FETCH_BALANCE, FETCH_BALANCE_SUCCESS, FETCH_BALANCE_FAILURE } = constants

// ------------------------------------
// Initial State
// ------------------------------------

/**
 * @typedef State
 * @property {boolean} isBalanceLoading Boolean indicating if balances are loading.
 * @property {string|null} walletBalance Total wallet balance.
 * @property {string|null} walletBalanceConfirmed Confirmed wallet balance.
 * @property {string|null} walletBalanceUnconfirmed Unconfirmed wallet balance.
 * @property {string|null} channelBalance Total channel balance.
 * @property {string|null} channelBalanceConfirmed Confirmed channel balance.
 * @property {string|null} channelBalancePending Pending channel balance.
 * @property {Error|null} fetchBalanceError Balance fetching error.
 */

/** @type {State} */
const initialState = {
  isBalanceLoading: false,
  walletBalance: null,
  walletBalanceConfirmed: null,
  walletBalanceUnconfirmed: null,
  channelBalance: null,
  channelBalanceConfirmed: null,
  channelBalancePending: null,
  fetchBalanceError: null,
}

// ------------------------------------
// Actions
// ------------------------------------

/**
 * fetchBalance - Fetch balances.
 *
 * @returns {(dispatch:Function) => Promise<void>} Thunk
 */
export const fetchBalance = () => async dispatch => {
  try {
    dispatch({ type: FETCH_BALANCE })
    const { walletBalance, channelBalance } = await grpc.services.Lightning.getBalance()
    dispatch({ type: FETCH_BALANCE_SUCCESS, walletBalance, channelBalance })
  } catch (error) {
    dispatch({ type: FETCH_BALANCE_FAILURE, error })
  }
}

// ------------------------------------
// Action Handlers
// ------------------------------------

const ACTION_HANDLERS = {
  [FETCH_BALANCE]: state => {
    state.isBalanceLoading = true
  },
  [FETCH_BALANCE_SUCCESS]: (state, { walletBalance, channelBalance }) => {
    state.isBalanceLoading = false
    state.walletBalance = walletBalance.totalBalance
    state.walletBalanceConfirmed = walletBalance.confirmedBalance
    state.walletBalanceUnconfirmed = walletBalance.unconfirmedBalance
    state.channelBalance = Coin(channelBalance.balance)
      .add(Coin(channelBalance.pendingOpenBalance))
      .toString()
    state.channelBalanceConfirmed = channelBalance.balance
    state.channelBalancePending = channelBalance.pendingOpenBalance
  },
  [FETCH_BALANCE_FAILURE]: (state, { error }) => {
    state.isBalanceLoading = false
    state.fetchBalanceError = error
  },
}

export default createReducer(initialState, ACTION_HANDLERS)
