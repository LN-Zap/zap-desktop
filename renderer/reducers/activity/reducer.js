import { send } from 'redux-electron-ipc'

import { getIntl } from '@zap/i18n'
import createReducer from '@zap/utils/createReducer'
import { mainLog } from '@zap/utils/log'
import { fetchBalance } from 'reducers/balance'
import { fetchChannels } from 'reducers/channels'
import { fetchInfo, infoSelectors } from 'reducers/info'
import { receiveInvoices } from 'reducers/invoice'
import { openModal, closeModal } from 'reducers/modal'
import { showError, showNotification } from 'reducers/notification'
import { receivePayments } from 'reducers/payment'
import { settingsSelectors } from 'reducers/settings'
import { receiveTransactions } from 'reducers/transaction'
import { walletSelectors } from 'reducers/wallet'

import * as constants from './constants'
import messages from './messages'
import { hasNextPage, isPageLoading } from './selectors'
import { createActivityPaginator } from './utils'

const {
  SET_ACTIVITY_MODAL,
  HIDE_ACTIVITY_MODAL,
  CHANGE_FILTER,
  UPDATE_SEARCH_TEXT,
  FETCH_ACTIVITY_HISTORY,
  FETCH_ACTIVITY_HISTORY_SUCCESS,
  FETCH_ACTIVITY_HISTORY_FAILURE,
  ADD_FILTER,
  SET_HAS_NEXT_PAGE,
  SET_ERROR_DIALOG_DETAILS,
  SET_PAGE_LOADING,
} = constants

// ------------------------------------
// Initial State
// ------------------------------------

/**
 * @typedef State
 * @property {Set<string>} filter Currently applied activity filter.
 * @property {object[]} filters List of supported activity filters.
 * @property {{itemType: string|null, itemId: string|null}} modal Activity item currently showing in activity modal.
 * @property {Error|null} errorDialogDetails Activity error currently showing in error details dialog.
 * @property {string|null} searchText Current activity search string.
 * @property {boolean} isActivityLoading Boolean indicating if activity is loading.
 * @property {Error|null} activityLoadingError Error from loading activity.
 * @property {boolean} hasNextPage Boolean indicating if there are more activity pages to fetch.
 * @property {boolean} isPageLoading Boolean indicating if the activity page is loading.
 */

/** @type {State} */
const initialState = {
  filter: new Set([...constants.defaultFilter]),
  filters: [
    { key: 'SENT_ACTIVITY' },
    { key: 'RECEIVED_ACTIVITY' },
    { key: 'PENDING_ACTIVITY' },
    { key: 'EXPIRED_ACTIVITY' },
    { key: 'INTERNAL_ACTIVITY' },
  ],
  modal: {
    itemType: null,
    itemId: null,
  },
  errorDialogDetails: null,
  searchText: null,
  isActivityLoading: false,
  activityLoadingError: null,
  hasNextPage: true,
  isPageLoading: false,
}

// ------------------------------------
// Helpers
// ------------------------------------

// activity paginator object. must be reset for each wallet login
/** @type {Function | null} */
let paginator = null

/**
 * getPaginator - Returns current activity paginator object. This acts as a singleton
 * and creates paginator if it's not initialized.
 *
 * @param {{
 *  invoiceHandler: function,
 *  paymentHandler: function,
 *  transactionHandler: function
 * }} handlers Pagination handlers
 * @returns {Function} Paginator
 */
export const getPaginator = handlers => {
  if (!paginator) {
    paginator = createActivityPaginator(handlers)
  }
  return paginator
}

// ------------------------------------
// Actions
// ------------------------------------

/**
 * setPageLoading - Set boolean indicating if page is in the process of loading.
 *
 * @param {boolean} isPageLoading Boolean indicating if page is loading
 * @returns {object} Action
 */
export const setPageLoading = isPageLoading => {
  return {
    type: SET_PAGE_LOADING,
    isPageLoading,
  }
}

/**
 * setErorDialogDetails - Set the error dialog details.
 *
 * @param {string} details Error details
 * @returns {object} Action
 */
export const setErorDialogDetails = details => {
  return {
    type: SET_ERROR_DIALOG_DETAILS,
    details,
  }
}

/**
 * changeFilter - Toggle specified activity filter.
 *
 * @param {...string} filterList Filter to apply
 * @returns {object} Action
 */
export function changeFilter(...filterList) {
  return {
    type: CHANGE_FILTER,
    filterList,
  }
}
/**
 * addFilter - Enable specified activity filters.
 *
 * @param {...string} filterList Filters to apply
 * @returns {object} Action
 */
export function addFilter(...filterList) {
  return {
    type: ADD_FILTER,
    filterList,
  }
}

/**
 * updateSearchText - Set the current activity search string.
 *
 * @param {string|null} searchText Search string to apply
 * @returns {object} Action
 */
export function updateSearchText(searchText = null) {
  return {
    type: UPDATE_SEARCH_TEXT,
    searchText,
  }
}

/**
 * saveInvoice - Initiates saving of invoice pdf.
 *
 * @param {object} options invoice options
 * @param {string} options.defaultFilename invoice title
 * @param {string} options.title invoice title
 * @param {string} options.subtitle invoice subtitle
 * @param {Array<object>} options.invoiceData invoice rows
 * @returns {(dispatch:Function) => Promise<void>} Thunk
 */
export const saveInvoice = ({
  defaultFilename,
  title,
  subtitle,
  invoiceData,
}) => async dispatch => {
  dispatch(send('saveInvoice', { defaultFilename, title, subtitle, invoiceData }))
}

/**
 * saveInvoiceFailure - Invoice save failed event.
 *
 * @returns {(dispatch:Function) => void} Thunk
 */
export const saveInvoiceFailure = () => dispatch => {
  dispatch(showError(getIntl().formatMessage(messages.activity_invoice_download_error)))
}

/**
 * saveInvoiceSuccess - Invoice save success event.
 *
 * @returns {(dispatch:Function) => void} Thunk
 */
export const saveInvoiceSuccess = () => dispatch => {
  dispatch(showNotification(getIntl().formatMessage(messages.activity_invoice_download_success)))
}

/**
 * setActivityModal - Show the activity modal with a given activity item.
 *
 * @param {string} itemType Item type
 * @param {string} itemId Item id
 * @returns {(dispatch:Function) => void} Thunk
 */
export const setActivityModal = (itemType, itemId) => dispatch => {
  dispatch({ type: SET_ACTIVITY_MODAL, itemType, itemId })
}

/**
 * showActivityModal - Show the activity modal with a given activity item.
 *
 * @param {string} itemType Item type
 * @param {string} itemId Item id
 * @returns {(dispatch:Function) => void} Thunk
 */
export const showActivityModal = (itemType, itemId) => dispatch => {
  dispatch(setActivityModal(itemType, itemId))
  dispatch(openModal('ACTIVITY_MODAL'))
}

/**
 * hideActivityModal - Hide the activity modal.
 *
 * @returns {(dispatch:Function) => void} Thunk
 */
export const hideActivityModal = () => dispatch => {
  dispatch({ type: HIDE_ACTIVITY_MODAL })
  dispatch(closeModal('ACTIVITY_MODAL'))
}

/**
 * resetActivity - Reset activity history.
 *
 * @returns {() => void} Thunk
 */
export const resetActivity = () => () => {
  paginator = null
}

/**
 * loadPage - Loads next activity page if it's available.
 *
 * @param {boolean} reload Boolean indicating wether to load first page.
 * @returns {(dispatch:Function, getState:Function) => Promise<void>} Thunk
 */
export const loadPage = (reload = false) => async (dispatch, getState) => {
  if (isPageLoading(getState())) {
    return
  }
  dispatch(setPageLoading(true))

  await dispatch(fetchInfo())
  const state = getState()
  const config = settingsSelectors.currentConfig(state)
  const blockHeight = infoSelectors.blockHeight(state)
  const activeWallet = walletSelectors.activeWallet(state)

  // Checks to see if the currently active wallet is the same one as when the page load started.
  const isStllActiveWallet = () => walletSelectors.activeWallet(getState()) === activeWallet

  // It's possible that the grpc response is received after the wallet has been disconnected.
  // See https://github.com/grpc/grpc-node/issues/1603
  // Ensure that the items received belong to the currently active wallet.
  const shouldUpdate = items => items && isStllActiveWallet()

  const handlers = {
    invoiceHandler: items => shouldUpdate(items) && dispatch(receiveInvoices(items)),
    paymentHandler: items => shouldUpdate(items) && dispatch(receivePayments(items)),
    transactionHandler: items => shouldUpdate(items) && dispatch(receiveTransactions(items)),
  }
  const thisPaginator = reload ? createActivityPaginator(handlers) : getPaginator(handlers)

  if (hasNextPage(getState())) {
    const { pageSize } = config.activity
    const { hasNextPage } = await thisPaginator(pageSize, blockHeight)
    if (!isStllActiveWallet()) {
      return
    }
    dispatch({ type: SET_HAS_NEXT_PAGE, value: hasNextPage })
  }

  dispatch(setPageLoading(false))
}

/**
 * reloadActivityHead - Reloads activity head.
 *
 * @returns {(dispatch:Function) => Promise<void>} Thunk
 */
export const reloadActivityHead = () => async dispatch => {
  mainLog.debug(`reloading activity pages`)
  await dispatch(loadPage(true))
}

/**
 * initActivityHistory - Load user activity history.
 *
 * @returns {(dispatch:Function) => void} Thunk
 */
export const initActivityHistory = () => async dispatch => {
  dispatch({ type: FETCH_ACTIVITY_HISTORY })
  try {
    await Promise.all([dispatch(loadPage()), dispatch(fetchChannels()), dispatch(fetchBalance())])
    dispatch({ type: FETCH_ACTIVITY_HISTORY_SUCCESS })
  } catch (error) {
    dispatch({ type: FETCH_ACTIVITY_HISTORY_FAILURE, error })
  }
}

/**
 * reloadActivityHistory - Reload activity history.
 *
 * @returns {(dispatch:Function) => void} Thunk
 */
export const reloadActivityHistory = () => async dispatch => {
  dispatch({ type: FETCH_ACTIVITY_HISTORY })
  try {
    await Promise.all([
      dispatch(reloadActivityHead()),
      dispatch(fetchChannels()),
      dispatch(fetchBalance()),
    ])
    dispatch({ type: FETCH_ACTIVITY_HISTORY_SUCCESS })
  } catch (error) {
    dispatch({ type: FETCH_ACTIVITY_HISTORY_FAILURE, error })
  }
}

// ------------------------------------
// Action Handlers
// ------------------------------------

const ACTION_HANDLERS = {
  [SET_ACTIVITY_MODAL]: (state, { itemType, itemId }) => {
    state.modal = { itemType, itemId }
  },
  [HIDE_ACTIVITY_MODAL]: state => {
    state.modal = { itemType: null, itemId: null }
  },
  [CHANGE_FILTER]: (state, { filterList }) => {
    const { filter } = state
    filterList.forEach(item => {
      if (filter.has(item)) {
        filter.delete(item)
      } else {
        filter.add(item)
      }
    })
  },
  [ADD_FILTER]: (state, { filterList }) => {
    const { filter } = state
    // No filters means all filters are active so no need to add anything
    if (filter.size > 0) {
      filterList.forEach(item => filter.add(item))
    }
  },
  [UPDATE_SEARCH_TEXT]: (state, { searchText }) => {
    state.searchText = searchText
  },
  [FETCH_ACTIVITY_HISTORY]: state => {
    state.isActivityLoading = true
  },
  [FETCH_ACTIVITY_HISTORY_SUCCESS]: state => {
    state.isActivityLoading = false
  },
  [FETCH_ACTIVITY_HISTORY_FAILURE]: (state, { error }) => {
    state.isActivityLoading = false
    state.activityLoadingError = error
  },
  [SET_HAS_NEXT_PAGE]: (state, { value }) => {
    state.hasNextPage = value
  },
  [SET_ERROR_DIALOG_DETAILS]: (state, { details }) => {
    state.errorDialogDetails = details
  },
  [SET_PAGE_LOADING]: (state, { isPageLoading }) => {
    state.isPageLoading = isPageLoading
  },
}

export default createReducer(initialState, ACTION_HANDLERS)
