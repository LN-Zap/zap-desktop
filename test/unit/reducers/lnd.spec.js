import reducer, {
  START_LND,
  START_LND_SUCCESS,
  START_LND_FAILURE,
  CLEAR_START_LND_ERROR,
  STOP_LND,
  STOP_LND_SUCCESS,
  CREATE_NEW_WALLET,
  CREATE_NEW_WALLET_SUCCESS,
  RECOVER_OLD_WALLET,
  RECOVER_OLD_WALLET_SUCCESS,
  RECOVER_OLD_WALLET_FAILURE,
  CLEAR_RECOVER_OLD_WALLET_ERROR,
  UNLOCK_WALLET,
  UNLOCK_WALLET_SUCCESS,
  UNLOCK_WALLET_FAILURE,
  FETCH_SEED,
  FETCH_SEED_FAILURE,
  FETCH_SEED_SUCCESS,
  CONNECT_GRPC,
  CONNECT_GRPC_SUCCESS,
  CONNECT_GRPC_FAILURE,
  LND_WALLET_UNLOCKER_GRPC_ACTIVE,
  LND_LIGHTNING_GRPC_ACTIVE,
  DISCONNECT_GRPC,
  DISCONNECT_GRPC_SUCCESS,
  DISCONNECT_GRPC_FAILURE,
} from 'reducers/lnd'

describe('reducers', () => {
  describe('neutrinoReducer', () => {
    it('should handle initial state', () => {
      expect(reducer(undefined, {})).toMatchSnapshot()
    })

    it('should handle START_LND', () => {
      const action = {
        type: START_LND,
        lndConfig: {
          type: 'local',
          chain: 'bitcoin',
          network: 'testnet',
        },
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle START_LND_SUCCESS', () => {
      const action = {
        type: START_LND_SUCCESS,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle START_LND_FAILURE', () => {
      const action = {
        type: START_LND_FAILURE,
        errors: {
          host: 'a host error',
          cert: 'a cert error',
          macaroon: 'a macaroon error',
        },
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle CLEAR_START_LND_ERROR', () => {
      const action = {
        type: CLEAR_START_LND_ERROR,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle STOP_LND', () => {
      const action = {
        type: STOP_LND,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle STOP_LND_SUCCESS', () => {
      const action = {
        type: STOP_LND_SUCCESS,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle CREATE_NEW_WALLET', () => {
      const action = {
        type: CREATE_NEW_WALLET,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle CREATE_NEW_WALLET_SUCCESS', () => {
      const action = {
        type: CREATE_NEW_WALLET_SUCCESS,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle RECOVER_OLD_WALLET', () => {
      const action = {
        type: RECOVER_OLD_WALLET,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle RECOVER_OLD_WALLET_SUCCESS', () => {
      const action = {
        type: RECOVER_OLD_WALLET_SUCCESS,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle RECOVER_OLD_WALLET_FAILURE', () => {
      const action = {
        type: RECOVER_OLD_WALLET_FAILURE,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle CLEAR_RECOVER_OLD_WALLET_ERROR', () => {
      const action = {
        type: CLEAR_RECOVER_OLD_WALLET_ERROR,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle UNLOCK_WALLET', () => {
      const action = {
        type: UNLOCK_WALLET,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle UNLOCK_WALLET_SUCCESS', () => {
      const action = {
        type: UNLOCK_WALLET_SUCCESS,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle UNLOCK_WALLET_FAILURE', () => {
      const action = {
        type: UNLOCK_WALLET_FAILURE,
        unlockWalletError: 'some error',
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle FETCH_SEED', () => {
      const action = {
        type: FETCH_SEED,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle FETCH_SEED_SUCCESS', () => {
      const action = {
        type: FETCH_SEED_SUCCESS,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle FETCH_SEED_FAILURE', () => {
      const action = {
        type: FETCH_SEED_FAILURE,
        error: 'some error',
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle CONNECT_GRPC', () => {
      const action = {
        type: CONNECT_GRPC,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle CONNECT_GRPC_SUCCESS', () => {
      const action = {
        type: CONNECT_GRPC_SUCCESS,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle CONNECT_GRPC_FAILURE', () => {
      const action = {
        type: CONNECT_GRPC_FAILURE,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle LND_WALLET_UNLOCKER_GRPC_ACTIVE', () => {
      const action = {
        type: LND_WALLET_UNLOCKER_GRPC_ACTIVE,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle LND_LIGHTNING_GRPC_ACTIVE', () => {
      const action = {
        type: LND_LIGHTNING_GRPC_ACTIVE,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle DISCONNECT_GRPC', () => {
      const action = {
        type: DISCONNECT_GRPC,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle DISCONNECT_GRPC_SUCCESS', () => {
      const action = {
        type: DISCONNECT_GRPC_SUCCESS,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })

    it('should handle DISCONNECT_GRPC_FAILURE', () => {
      const action = {
        type: DISCONNECT_GRPC_FAILURE,
      }
      expect(reducer({}, action)).toMatchSnapshot()
    })
  })
})
