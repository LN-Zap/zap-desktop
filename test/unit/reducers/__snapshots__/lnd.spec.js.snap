// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers neutrinoReducer should handle CLEAR_RECOVER_OLD_WALLET_ERROR 1`] = `
Object {
  "walletRecoveryError": null,
}
`;

exports[`reducers neutrinoReducer should handle CLEAR_START_LND_ERROR 1`] = `
Object {
  "startLndError": null,
}
`;

exports[`reducers neutrinoReducer should handle CONNECT_GRPC 1`] = `
Object {
  "isStartingGrpc": true,
}
`;

exports[`reducers neutrinoReducer should handle CONNECT_GRPC_FAILURE 1`] = `
Object {
  "isStartingGrpc": false,
}
`;

exports[`reducers neutrinoReducer should handle CONNECT_GRPC_SUCCESS 1`] = `
Object {
  "isStartingGrpc": false,
}
`;

exports[`reducers neutrinoReducer should handle CREATE_NEW_WALLET 1`] = `
Object {
  "isCreatingNewWallet": true,
}
`;

exports[`reducers neutrinoReducer should handle CREATE_NEW_WALLET_SUCCESS 1`] = `
Object {
  "isCreatingNewWallet": false,
}
`;

exports[`reducers neutrinoReducer should handle DISCONNECT_GRPC 1`] = `
Object {
  "isStoppingGrpc": true,
}
`;

exports[`reducers neutrinoReducer should handle DISCONNECT_GRPC_FAILURE 1`] = `
Object {
  "isStoppingGrpc": false,
}
`;

exports[`reducers neutrinoReducer should handle DISCONNECT_GRPC_SUCCESS 1`] = `
Object {
  "isLightningGrpcActive": false,
  "isStoppingGrpc": false,
  "isWalletUnlockerGrpcActive": false,
}
`;

exports[`reducers neutrinoReducer should handle FETCH_SEED 1`] = `
Object {
  "isFetchingSeed": true,
}
`;

exports[`reducers neutrinoReducer should handle FETCH_SEED_FAILURE 1`] = `
Object {
  "fetchSeedError": "some error",
  "isFetchingSeed": false,
}
`;

exports[`reducers neutrinoReducer should handle FETCH_SEED_SUCCESS 1`] = `
Object {
  "fetchSeedError": null,
  "isFetchingSeed": false,
}
`;

exports[`reducers neutrinoReducer should handle LND_LIGHTNING_GRPC_ACTIVE 1`] = `
Object {
  "isLightningGrpcActive": true,
  "isWalletUnlockerGrpcActive": false,
}
`;

exports[`reducers neutrinoReducer should handle LND_WALLET_UNLOCKER_GRPC_ACTIVE 1`] = `
Object {
  "isLightningGrpcActive": false,
  "isWalletUnlockerGrpcActive": true,
}
`;

exports[`reducers neutrinoReducer should handle RECOVER_OLD_WALLET 1`] = `
Object {
  "isRecoveringWallet": true,
  "walletRecoveryError": null,
}
`;

exports[`reducers neutrinoReducer should handle RECOVER_OLD_WALLET_FAILURE 1`] = `
Object {
  "isRecoveringWallet": false,
  "walletRecoveryError": undefined,
}
`;

exports[`reducers neutrinoReducer should handle RECOVER_OLD_WALLET_SUCCESS 1`] = `
Object {
  "isRecoveringWallet": false,
  "walletRecoveryError": null,
}
`;

exports[`reducers neutrinoReducer should handle START_LND 1`] = `
Object {
  "isStartingLnd": true,
  "lndConfig": Object {
    "chain": "bitcoin",
    "network": "testnet",
    "type": "local",
  },
}
`;

exports[`reducers neutrinoReducer should handle START_LND_FAILURE 1`] = `
Object {
  "isStartingLnd": false,
  "startLndError": Object {
    "cert": "a cert error",
    "host": "a host error",
    "macaroon": "a macaroon error",
  },
}
`;

exports[`reducers neutrinoReducer should handle START_LND_SUCCESS 1`] = `
Object {
  "isLndActive": true,
  "isStartingLnd": false,
}
`;

exports[`reducers neutrinoReducer should handle STOP_LND 1`] = `
Object {
  "isStoppingLnd": true,
}
`;

exports[`reducers neutrinoReducer should handle STOP_LND_SUCCESS 1`] = `
Object {
  "isLndActive": false,
  "isStoppingLnd": false,
}
`;

exports[`reducers neutrinoReducer should handle UNLOCK_WALLET 1`] = `
Object {
  "isUnlockingWallet": true,
}
`;

exports[`reducers neutrinoReducer should handle UNLOCK_WALLET_FAILURE 1`] = `
Object {
  "isUnlockingWallet": false,
  "unlockWalletError": "some error",
}
`;

exports[`reducers neutrinoReducer should handle UNLOCK_WALLET_SUCCESS 1`] = `
Object {
  "isUnlockingWallet": false,
  "unlockWalletError": null,
}
`;

exports[`reducers neutrinoReducer should handle initial state 1`] = `
Object {
  "fetchSeedError": null,
  "isCreatingNewWallet": false,
  "isFetchingSeed": false,
  "isLightningGrpcActive": false,
  "isLndActive": false,
  "isRecoveringWallet": false,
  "isStartingGrpc": false,
  "isStartingLnd": false,
  "isStoppingLnd": false,
  "isUnlockingWallet": false,
  "isWalletUnlockerGrpcActive": false,
  "lndConfig": Object {},
  "startLndError": null,
  "unlockWalletError": null,
  "walletRecoveryError": null,
}
`;
